tests/integration/targets/nxos_file_copy/tests/cli/sanity.yaml

```diff
@@ -22,6 +25,23 @@
     feature: scp-server
     state: enabled
 
+- name: Setup - Turn on feature sftp-server
+  cisco.nxos.nxos_feature:
+    feature: sftp-server
+    state: enabled
+
+- name: Setup - Backup admin password
```

Comment:
Why can we not use the `temp_user` created in tasks/main.yml like we did in the already existing file_pull (with scp) test?

Reply:
Hi,
To test the additional functionality I am enabling the sftp-server feature of NX OS. Unfortunately only the admin can use SFTP to access the NX OS. See chapter "Guidelines and Limitations for SSH and Telnet" on the page [Cisco Nexus 9000 Series NX-OS Security Configuration Guide, Release 7.x](https://www.cisco.com/c/en/us/td/docs/switches/datacenter/nexus9000/sw/7-x/security/configuration/guide/b_Cisco_Nexus_9000_Series_NX-OS_Security_Configuration_Guide_7x/b_Cisco_Nexus_9000_Series_NX-OS_Security_Configuration_Guide_7x_chapter_0111.html)
> When the SFTP server is enabled, only the admin user can use SFTP to access the device.

That's why I am not able to use the `tempuser`. 
I wasn't able to find a way to enable another file server like http or tftp on the NX-OS to avoid using the admin user. I think only scp or sftp are supported as a server service on NX-OS.

Reply:
Changing the `admin` user password midway will cause the succeeding tasks to error out with `Authentication Failures`. The appliances in our CI environment uses key-based auth and hence, we don't see this issue there. However, for anyone running the test suite locally (with `ansible_user: admin`), it would fail.

I wanted to try out a couple of things to get past this issue, but unfortunately pulling a file from the local SFTP server (with `admin` user) is not working for me. Trying the same thing using SCP, works. Here's what's happening:
```
NXOS-9K-935# show file bootflash:///test_destination_file
TEST FILE

NXOS-9K-935# copy scp://admin@10.0.2.15/test_destination_file bootflash:/dir1/dir2/dir3/test_destination_file_copy vrf management
Outbound-ReKey for 10.0.2.15:22
Inbound-ReKey for 10.0.2.15:22
User Access Verification
Password:
test_destination_file                                                                                               100%   10     3.2KB/s   00:00
Copy complete, now saving to disk (please wait)...
Copy complete.

NXOS-9K-935# copy sftp://admin@10.0.2.15/test_destination_file bootflash:/dir1/dir2/dir3/test_destination_file_copy vrf management
Warning: There is already a file existing with this name. Do you want to overwrite (y/n)?[n] y

Outbound-ReKey for 10.0.2.15:22
Inbound-ReKey for 10.0.2.15:22
User Access Verification
Password:
sftp> progress
Progress meter enabled
sftp> get   /test_destination_file  /bootflash/dir1/dir2/dir3/test_destination_file_copy
File "/test_destination_file" not found.
Copy failed. Removing file dir1/dir2/dir3/test_destination_file_copy
```

I was trying to check if using a different user with `network-admin` privilege would work. We can make `temp_user` an admin by adding `role: network-admin` in the task that creates it. Since I'm not able to test this, could you try it out and see if it works? If not, we'd have to resort to adding an unit test to cover this use case.

Reply:
I can understand your concerns regarding changing the admin password. I didn't think about that.

Are you using in your tests with SFTP NX-OS 7 or 9? If you have used a NX-OS version greater than 9, then you need to add "bootflash" in your sftp path. e.g. 
`copy sftp://admin@10.0.2.15/bootflash/test_destination_file bootflash:/dir1/dir2/dir3/test_destination_file_copy vrf management`
That was the reason why I am setting the following fact:
```
- set_fact:
    sftp_root_dir: "{% if major_version is version('9.2', 'ge') %}/bootflash{% else %}/{% endif %}"
```


I have assigned the network-admin role to temp_user, but unfortunately it doesn't work. See output:
```
Nexus9000v(config)# sh version | i "NXOS: version"
  NXOS: version 7.0(3)I7(5)
Nexus9000v(config)# username temp_user password xxxxxxx role network-admin
Nexus9000v(config)# copy sftp://temp_user@10.0.2.15/virtual-instance.conf bootflash:bla.txt vrf management

User Access Verification
Password:
Connection closed
Nexus9000v(config)# sh logg last 1
2020 Sep 29 19:00:28 Nexus9000v %DAEMON-3-SYSTEM_MSG: error: sftp allowed for user admin only - dcos_sshd[17390]
Nexus9000v(config)#
Nexus9000v(config)# copy sftp://admin@10.0.2.15/virtual-instance.conf bootflash:bla.txt vrf management

User Access Verification
Password:
sftp> progress
Progress meter enabled
sftp> get   /virtual-instance.conf  /bootflash/bla.txt
/virtual-instance.conf                                                                                                                                                                                     100%   59     0.1KB/s   00:00
sftp> exit
Copy complete, now saving to disk (please wait)...
Copy complete.
Nexus9000v(config)#
```
But I found another way how to handle the admin password. If someone uses the admin password to run the integration test, the test case will use the admin password from the ansible_password variable and skip all the tasks, which manipulate the admin user. If anyone uses another account, it will change the admin password temporarily. I have run successfully locally the integration tests with admin account and with another user account. Check my last commit.
What do you think? Is this a possible way.

Reply:
@thushjandan You are right, adding "bootflash" to the path fixed it. But now I can see that a test user with network-admin privilege works.

```
NXOS-9K-935# show version | i "NXOS: version"
 NXOS: version 9.3(5)

NXOS-9K-935# show user-account test_user
user:test_user
        this user account has no expiry date
        roles:network-admin

NXOS-9K-935# copy sftp://test_user@10.0.2.15/bootflash/test_destination_file bootflash:/dir1/dir2/dir3/test_destination_file_copy vrf management
Warning: There is already a file existing with this name. Do you want to overwrite (y/n)?[n] y

Outbound-ReKey for 10.0.2.15:22
Inbound-ReKey for 10.0.2.15:22
User Access Verification
Password:
sftp> progress
Progress meter enabled
sftp> get   /bootflash/test_destination_file  /bootflash/dir1/dir2/dir3/test_destination_file_copy
/bootflash/test_destination_file                                                                                     100%   10     1.1KB/s   00:00
sftp> exit
Copy complete, now saving to disk (please wait)...
Copy complete.
```

Reply:
Okay, it seems that this^^ is not universally true. I tested this against a N9Kv running an older NX-OS (which is what we have in CI) and I was able to reproduce what you're facing:
```
NXOS-9K-I5# show ver | i "NXOS: version"
  NXOS: version 7.0(3)I5(1)

NXOS-9K-I5# show user-account test_user
user:test_user
        this user account has no expiry date
        roles:network-admin

NXOS-9K-I5# copy sftp://test_user@10.0.2.15/virtual-instance.conf bootflash:test vrf management
Password:
Connection closed

NXOS-9K-I5# show logg last 1
2020 Sep 30 07:40:02 NXOS-9K-I5 %DAEMON-3-SYSTEM_MSG: error: sftp allowed for user admin only - sshd[27898]
```

Reply:
The changes in the latest commit look perfectly reasonable to me. Thanks a lot for helping with this PR, I really appreciate it! We'll merge this now and it should be shipped with the upcoming v1.2.0 release.

Thanks again and looking forward to more PRs from you. :)